Realm/Tests/NotificationTests.m

```diff
@@ -0,0 +1,633 @@
+        [array addObject:[IntObject createInRealm:realm withValue:@[@3]]];
+        [realm deleteObject:[IntObject allObjectsInRealm:realm].lastObject];
+    });
+}
+
+- (void)testInsertObjectMatchingQuery {
+    ExpectChange(self, @[], @[@4], @[], ^(RLMRealm *realm) {
+        RLMArray *array = [[[ArrayPropertyObject allObjectsInRealm:realm] firstObject] intArray];
+        [array addObject:[IntObject createInRealm:realm withValue:@[@3]]];
+    });
+}
+
+- (void)testInsertObjectNotMatchingQuery {
+    ExpectNoChange(self, ^(RLMRealm *realm) {
+        RLMArray *array = [[[ArrayPropertyObject allObjectsInRealm:realm] firstObject] intArray];
+        [array addObject:[IntObject createInRealm:realm withValue:@[@5]]];
+    });
+}
+
+- (void)testInsertBothMatchingAndNonMatching {
+    ExpectChange(self, @[], @[@4], @[], ^(RLMRealm *realm) {
+        RLMArray *array = [[[ArrayPropertyObject allObjectsInRealm:realm] firstObject] intArray];
+        [array addObject:[IntObject createInRealm:realm withValue:@[@5]]];
+        [array addObject:[IntObject createInRealm:realm withValue:@[@3]]];
+    });
+}
+
+- (void)testInsertMultipleMatching {
+    ExpectChange(self, @[], @[@4, @5], @[], ^(RLMRealm *realm) {
+        RLMArray *array = [[[ArrayPropertyObject allObjectsInRealm:realm] firstObject] intArray];
+        [array addObject:[IntObject createInRealm:realm withValue:@[@5]]];
+        [array addObject:[IntObject createInRealm:realm withValue:@[@3]]];
+        [array addObject:[IntObject createInRealm:realm withValue:@[@5]]];
+        [array addObject:[IntObject createInRealm:realm withValue:@[@2]]];
+    });
+}
+
+- (void)testInsertAtIndex {
+    ExpectChange(self, @[], @[@0], @[], ^(RLMRealm *realm) {
+        RLMArray *array = [[[ArrayPropertyObject allObjectsInRealm:realm] firstObject] intArray];
+        IntObject *io = [IntObject createInRealm:realm withValue:@[@3]];
+        [array insertObject:io atIndex:0];
+    });
+
+    ExpectChange(self, @[], @[@0], @[], ^(RLMRealm *realm) {
+        RLMArray *array = [[[ArrayPropertyObject allObjectsInRealm:realm] firstObject] intArray];
+        IntObject *io = [IntObject createInRealm:realm withValue:@[@3]];
+        [array insertObject:io atIndex:1];
+    });
+
+    ExpectChange(self, @[], @[@1], @[], ^(RLMRealm *realm) {
+        RLMArray *array = [[[ArrayPropertyObject allObjectsInRealm:realm] firstObject] intArray];
+        IntObject *io = [IntObject createInRealm:realm withValue:@[@3]];
+        [array insertObject:io atIndex:2];
+    });
+
+    ExpectNoChange(self, ^(RLMRealm *realm) {
+        RLMArray *array = [[[ArrayPropertyObject allObjectsInRealm:realm] firstObject] intArray];
+        IntObject *io = [IntObject createInRealm:realm withValue:@[@5]];
+        [array insertObject:io atIndex:2];
+    });
+}
+
+- (void)testExchangeObjects {
+    // adjacent swap: one move, since second is redundant
+//    ExpectChange(self, @[@1, @0], @[], @[], ^(RLMRealm *realm) {
```

Comment:
I never implemented the non-naive version of swap handling that produces nice diffs.